cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasConfiguration.qll

```diff
@@ -146,3 +146,8 @@ class DynamicAllocation extends Allocation, TDynamicAllocation {
 }
 
 predicate phaseNeedsSoundEscapeAnalysis() { none() }
+
+UnaliasedSsa::Allocation getOldAllocation(VariableAllocation allocation) {
+  UnaliasedSsa::canReuseSsaForVariable(allocation.getIRVariable()) and
```

Comment:
What exactly does `canReuseSsaForVariable` do? I've read the comment on the predicate, but I don't really undertstand why this is important here.

Reply:
As you know we do 2 SSA analyses as part of IR construction:
- One when we construct the unaliased IR
- One when we construct the aliased (i.e., "final") IR

The unaliased IR computed SSA information for those variables that never alias with anything else (hence the name), and the aliased IR computes everything else. Thus, the aliased IR needs to know which variables it is in charge of computing SSA for, and which variables it can just reuse SSA the unaliased SSA's computations for.

So, in this predicate, when we're asking for the allocation produced from by unaliased SSA (i.e., a `UnaliasedSsa::Allocation`) for a given aliased allocation (i.e., a `VariableAllocation`) we only return a result if the alias information for the aliased allocation can actually be reused from the unaliased SSA analysis.

And the reason for why we need the `getOldAllocation` predicate is that we need it for the base case in the new `propagatedFromAllocationBase` predicate in `AliasAnalysis.qll`:
```ql
private predicate propagatedFromAllocationBase(Operand operand, Configuration::Allocation allocation) {
  consumedAsAddressOperand(operand) and
  (
    not exists(Configuration::getOldAllocation(allocation)) and
    operand.getDef() = allocation.getABaseInstruction()
    or
    exists(Operand address |
      operandIsPropagated(address, _, operand.getDef()) and
      propagatedFromAllocationBase(address, allocation)
    )
  )
}
```
The main predicate that I'm adding in this PR is `operandConsumesEscaped` (which captures the situation expressed in the PR description). And we only want to reason about allocations that haven't already been resolved in the unaliased stage. So we want to insert the `not exists(Configuration::getOldAllocation(allocation))` at the earliest point to restrict the size of the recursion, and the base case of this recursion is the earliest possible place to put it 😄

Reply:
Thanks. That helps a lot.